// @conform-to/react/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var dom = require('@conform-to/dom');
var context = require('./context.js');
var hooks = require('./hooks.js');
var integrations = require('./integrations.js');
var helpers = require('./helpers.js');



Object.defineProperty(exports, 'parse', {
	enumerable: true,
	get: function () { return dom.parse; }
});
exports.FormProvider = context.FormProvider;
exports.FormStateInput = context.FormStateInput;
exports.useField = hooks.useField;
exports.useForm = hooks.useForm;
exports.useFormMetadata = hooks.useFormMetadata;
exports.unstable_Control = integrations.Control;
exports.unstable_useControl = integrations.useControl;
exports.useInputControl = integrations.useInputControl;
exports.getCollectionProps = helpers.getCollectionProps;
exports.getFieldsetProps = helpers.getFieldsetProps;
exports.getFormProps = helpers.getFormProps;
exports.getInputProps = helpers.getInputProps;
exports.getSelectProps = helpers.getSelectProps;
exports.getTextareaProps = helpers.getTextareaProps;
